bfsscp/bfsscp.go

```diff
@@ -0,0 +1,440 @@
+		*config = *cfg
+	}
+	if err := config.norm(); err != nil {
+		return nil, err
+	}
+
+	sshConfig := &ssh.ClientConfig{
+		User: cfg.Username,
+		Auth: []ssh.AuthMethod{
+			ssh.Password(cfg.Password),
+		},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	conn, err := ssh.Dial("tcp", address, sshConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := sftp.NewClient(conn)
+	if err != nil {
+		return nil, multierr.Combine(conn.Close(), err)
+	}
+
+	return &bucket{
+		conn:   conn,
+		client: client,
+		config: config,
+	}, nil
+}
+
+func (b *bucket) withPrefix(name string) string {
+	if b.config.Prefix == "" {
+		return name
+	}
+	return internal.WithinNamespace(b.config.Prefix, name)
+}
+
+// Glob implements bfs.Bucket.
+func (b *bucket) Glob(ctx context.Context, pattern string) (bfs.Iterator, error) {
+	// quick sanity check
+	if _, err := doublestar.Match(pattern, ""); err != nil {
+		return nil, err
+	}
+
+	// Need to walk the tree for global pattern
+	if strings.Contains(pattern, "**") {
+		// create a new walker iterator
+		return &walkerIterator{
+			w:            b.client.Walk(b.withPrefix("/")),
+			infoIterator: infoIterator{ctx: ctx},
+		}, nil
+	}
+
+	return newMatchesIterator(ctx, b.client, b.withPrefix(pattern))
+}
+
+// Head implements bfs.Bucket.
+func (b *bucket) Head(ctx context.Context, name string) (*bfs.MetaInfo, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	info, err := b.client.Stat(b.withPrefix(name))
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	return &bfs.MetaInfo{
+		Name:    name,
+		Size:    info.Size(),
+		ModTime: info.ModTime(),
+	}, nil
+}
+
+// Open implements bfs.Bucket.
+func (b *bucket) Open(ctx context.Context, name string) (bfs.Reader, error) {
+	file, err := b.client.Open(b.withPrefix(name))
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	info, err := file.Stat()
+	if err != nil {
+		_ = file.Close()
+		return nil, normError(err)
+	}
+
+	// Use file size to ensure limited read and avoid EOFs
+	return &reader{File: file, r: io.LimitReader(file, info.Size()), ctx: ctx}, nil
+}
+
+// Create implements bfs.Bucket.
+func (b *bucket) Create(ctx context.Context, name string, opts *bfs.WriteOptions) (bfs.Writer, error) {
+	f, err := ioutil.TempFile(b.config.TempDir, "bfs-scp")
+	if err != nil {
+		return nil, err
+	}
+
+	return &writer{
+		File:   f,
+		bucket: b,
+		ctx:    ctx,
+		name:   name,
+		opts:   opts,
+	}, nil
+}
+
+// Remove implements bfs.Bucket.
+func (b *bucket) Remove(ctx context.Context, name string) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	err := normError(b.client.Remove(b.withPrefix(name)))
+	if err != nil && err != bfs.ErrNotFound {
+		return err
+	}
+	return nil
+}
+
+// Close implements bfs.Bucket.
+func (b *bucket) Close() error {
+	return multierr.Combine(b.client.Close(), b.conn.Close())
+}
+
+// --------------------------------------------------------
+
+type writer struct {
+	*os.File
```

Comment:
sorry, stupid question, but there doesn't seem to be any benefit of embedding the File, or is there?

Reply:
Not to duplicate `Write` method?

Reply:
Actually, I'd suggest to make it a prop and call it like `tmp` - it's slightly surprising to see stuff like `w.Name()` and `w.name` down the code.